src/mpi/coll/transports/gentran/tsp_gentran.c

```diff
@@ -92,25 +92,6 @@ void MPII_Genutil_sched_free(MPII_Genutil_sched_t * sched)
     MPL_free(sched);
 }
 
-int MPII_Genutil_sched_new_type(MPIR_TSP_sched_t s, MPII_Genutil_sched_issue_fn issue_fn,
-                                MPII_Genutil_sched_complete_fn complete_fn,
-                                MPII_Genutil_sched_free_fn free_fn)
-{
-    MPII_Genutil_sched_t *sched = s;
-    MPII_Genutil_vtx_type_t *newtype;
-    int type_id;
-
-    type_id = MPII_GENUTIL_VTX_KIND__LAST + utarray_len(&sched->generic_types) + 1;
-    utarray_extend_back(&sched->generic_types, MPL_MEM_COLL);
-    newtype = (MPII_Genutil_vtx_type_t *) utarray_back(&sched->generic_types);
-    newtype->id = type_id;
-    newtype->issue_fn = issue_fn;
-    newtype->complete_fn = complete_fn;
-    newtype->free_fn = free_fn;
-
-    return type_id;
-}
-
 int MPII_Genutil_sched_generic(int type_id, void *data,
```

Comment:
can we undo this commit?
The new_type interface is actually used by us. This interface is important so that we can add new node type in the schedule tree.

Reply:
I see. So we have to expose the internal gentran data structure?

Reply:
I don't see how it really exposes the internal data structure. But I agree it is specific to Gentran.

Reply:
The function pointers would need the vertex data struct definitions, right?

Reply:
You are probably right. anyway what is the issue with that? It is just part of the gentran API.

Reply:
Oh, I was trying to make the TSP interface abstract. Abstract interfaces are easier to maintain and extend. Since there is code breaking that abstraction, I guess we'll just deal with it.

I am a bit curious how did it happen. Isn't the whole design point originally was to keep TSP interface abstract so we could swap the gentran with alternate transport?

Reply:
Swappable TSP interface probably never worked. The latest idea is to make gentran more general and extendable with this new interface so that it works for new algorithms.

Reply:
ok

Reply:
@zhenggb72 In the prototypes of callback functions:
https://github.com/zhenggb72/mpich/blob/af753cae3b43e5e12a4030f6841328b1d238c917/src/mpi/coll/transports/gentran/tsp_gentran_types.h#L126-L128
Is it sufficient to pass in the `void *data` instead of `MPII_Genutil_vtx_t * vtxp`?

Reply:
@zhenggb72 The mentioned commit has been updated in https://github.com/pmodels/mpich/pull/5313/commits/f209fa76cf036aa244368c15322b2bb47f32431d. Please review.